File: kirun-js/src/engine/dsl/parser/ast/FunctionCallNode.ts
import { SourceLocation } from '../../lexer/DSLToken';
import { ArgumentNode } from './ArgumentNode';
import { ASTNode } from './ASTNode';

/**
 * Function call node
 * Example: System.Array.InsertLast(source = Context.a, element = 10)
 */
export class FunctionCallNode extends ASTNode {
    constructor(
        public namespace: string,
        public name: string,
        public argumentsMap: Map<string, ArgumentNode>,
        location: SourceLocation,
    ) {
        super('FunctionCall', location);
    }

    public toJSON(): any {
        return {
            type: this.type,
            namespace: this.namespace,
            name: this.name,
            arguments: Object.fromEntries(
                Array.from(this.argumentsMap.entries()).map(([key, arg]) => [key, arg.toJSON()]),
            ),
        };
    }
}

File: kirun-py/src/kirun_py/dsl/parser/ast/function_call_node.py
from __future__ import annotations

from typing import Any, Dict

from kirun_py.dsl.lexer.dsl_token import SourceLocation
from kirun_py.dsl.parser.ast.argument_node import ArgumentNode
from kirun_py.dsl.parser.ast.ast_node import ASTNode


class FunctionCallNode(ASTNode):
    """
    Function call node.
    Example: System.Array.InsertLast(source = Context.a, element = 10)
    """

    def __init__(
        self,
        namespace: str,
        name: str,
        arguments_map: Dict[str, ArgumentNode],
        location: SourceLocation,
    ) -> None:
        super().__init__('FunctionCall', location)
        self.namespace = namespace
        self.name = name
        self.arguments_map = arguments_map

    def to_json(self) -> Any:
        return {
            'type': self.type,
            'namespace': self.namespace,
            'name': self.name,
            'arguments': {key: arg.to_json() for key, arg in self.arguments_map.items()},
        }
